test/Microsoft.VisualStudio.LanguageServices.Razor.Test/ProjectSystem/DefaultProjectSnapshotManagerTest.cs

```diff
@@ -0,0 +1,815 @@
+
+            var project = WorkspaceProject.WithAssemblyName("Test1"); // Simulate a project change
+            ProjectManager.WorkspaceProjectChanged(project);
+            ProjectManager.Reset();
+
+            // Generate the update
+            var snapshot = ProjectManager.GetSnapshot(WorkspaceProject);
+            var updateContext = snapshot.CreateUpdateContext();
+
+            // Act
+            ProjectManager.ProjectUpdated(updateContext);
+
+            // Assert
+            snapshot = ProjectManager.GetSnapshot(project);
+            Assert.False(snapshot.IsDirty);
+
+            Assert.True(ProjectManager.ListenersNotified);
+            Assert.False(ProjectManager.WorkerStarted);
+        }
+
+        [ForegroundFact]
+        public void ProjectUpdated_WhenHostProjectChanged_StillDirty_WithSignificantChanges_NotifiesListeners_AndStartsBackgroundWorker()
+        {
+            // Arrange
+            ProjectManager.HostProjectAdded(HostProject);
+            ProjectManager.WorkspaceProjectAdded(WorkspaceProject);
+            ProjectManager.Reset();
+
+            // Generate the update
+            var snapshot = ProjectManager.GetSnapshot(HostProject);
+            var updateContext = snapshot.CreateUpdateContext();
+
+            var project = new HostProject(HostProject.FilePath, FallbackRazorConfiguration.MVC_1_0); // Simulate a project change
+            ProjectManager.HostProjectChanged(project);
+            ProjectManager.Reset();
+
+            // Act
+            ProjectManager.ProjectUpdated(updateContext);
+
+            // Assert
+            snapshot = ProjectManager.GetSnapshot(project);
+            Assert.True(snapshot.IsDirty);
+
+            Assert.True(ProjectManager.ListenersNotified);
+            Assert.True(ProjectManager.WorkerStarted);
+        }
+
+        [ForegroundFact]
+        public void WorkspaceProjectChanged_BackgroundUpdate_StillDirty_WithSignificantChanges_NotifiesListeners_AndStartsBackgroundWorker()
+        {
+            // Arrange
+            ProjectManager.HostProjectAdded(HostProject);
+            ProjectManager.WorkspaceProjectAdded(WorkspaceProject);
+            ProjectManager.Reset();
+
+            // Generate the update
+            var snapshot = ProjectManager.GetSnapshot(WorkspaceProject);
+            var updateContext = snapshot.CreateUpdateContext();
+
+            var project = WorkspaceProject.WithAssemblyName("Test1"); // Simulate a project change
+            ProjectManager.WorkspaceProjectChanged(project);
+            ProjectManager.Reset();
+
+            // Act
+            ProjectManager.ProjectUpdated(updateContext);
+
+            // Assert
+            snapshot = ProjectManager.GetSnapshot(project);
+            Assert.True(snapshot.IsDirty);
+
+            Assert.True(ProjectManager.ListenersNotified);
+            Assert.True(ProjectManager.WorkerStarted);
+        }
+
+        [Fact(Skip = "We no longer have any background-computed state")]
```

Comment:
Couldn't unskip the skipped project snapshot manager tests yet. I did not add the concept of a project is dirty but to ignore that dirtiness based on a hueristic.